File: spring-boot-starter-breuninger-validation/src/main/java/com/breuninger/boot/validation/validators/SafeIdValidator.java
package com.breuninger.boot.validation.validators;

import java.lang.annotation.Annotation;
import java.util.regex.Pattern;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class SafeIdValidator implements ConstraintValidator<Annotation, String> {

  private static final Pattern IdPattern = Pattern.compile("[a-zA-Z0-9\\-_]*");

  @Override
  public void initialize(final Annotation annotation) {
  }

  @Override
  public boolean isValid(final String id, final ConstraintValidatorContext context) {
    if (id == null) {
      return true;
    }
    return IdPattern.matcher(id).matches();
  }
}

File: validation/src/main/kotlin/com/breuninger/boot/validation/constraintvalidators/SafeIdValidator.kt
package com.breuninger.boot.validation.constraintvalidators

import com.breuninger.boot.validation.constraints.SafeId
import java.util.regex.Pattern
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class SafeIdValidator : ConstraintValidator<SafeId, String> {

  companion object {

    private val SAFE_ID_PATTERN = Pattern.compile("[a-zA-Z0-9\\-_]*")
  }

  override fun initialize(safeId: SafeId?) {
  }

  override fun isValid(id: String?, context: ConstraintValidatorContext) =
    id?.let { SAFE_ID_PATTERN.matcher(it).matches() } ?: true
}
